packages/embeds/embed-core/src/embed.ts

```diff
@@ -1037,41 +1020,99 @@ class CalApi {
       containerEl
     );
 
-    const previousEmbedRenderStartTime = this.cal.embedRenderStartTime;
-    const previousEmbedConfig = this.cal.embedConfig;
+    let enrichedPrerenderOptions;
+    if (isPrerendering) {
+      const preparationResult = this.cal.prepareForPrerender({
+        calLink,
+        calOrigin,
+        previousEmbedRenderStartTime,
+        prerenderOptions,
+        isHeadlessRouterPath,
+      });
 
+      if (!preparationResult) {
+        log(`Prevented unnecessary repeat prerender for ${calLink}`);
+        return {
+          status: "prerender-prevented",
+        };
+      }
+
+      enrichedPrerenderOptions = preparationResult.enrichedPrerenderOptions;
+      enrichedConfig = buildConfigWithPrerenderRelatedFields({
+        config: configWithGuestKeyAndColorScheme,
+        isHeadlessRouterPath,
+        backgroundSlotsFetch: !!this.prerenderOptions?.backgroundSlotsFetch,
+      });
+    } else {
+      this.cal.isPrerendering = false;
+      enrichedConfig = configWithGuestKeyAndColorScheme;
+    }
+
+    // `this.modalUid` is set in non-preload case(Temporarily not being-set)
+    // `this.prerenderedModalUid` is set for a modal created through "prerender"
+    const reusableModalUid = this.modalUid || this.prerenderedModalUid;
+    const uid = reusableModalUid || String(Date.now());
+    // Means whether there is already an attempt to use the prerendered modal
+    const isConnectionInitiated = !!(this.modalUid && this.prerenderedModalUid);
     const embedRenderStartTime = Date.now();
-    this.cal.embedRenderStartTime = embedRenderStartTime;
-    this.cal.embedConfig = configWithGuestKeyAndColorScheme;
 
-    const existingModalEl = document.querySelector(`cal-modal-box[uid="${uid}"]`);
+    const stateData = {
+      embedConfig: enrichedConfig,
+      previousEmbedConfig,
+      embedRenderStartTime,
+      previousEmbedRenderStartTime,
+      isConnectionInitiated,
+      // If prerenderOptions are available, use them. Otherwise use the prerenderOptions from the previous call to modal for prerender
+      prerenderOptions: enrichedPrerenderOptions ?? this.prerenderOptions ?? null,
+    };
 
+    this.cal.setModalRenderStartVariables({
+      embedConfig: enrichedConfig,
+      embedRenderStartTime,
+    });
+
+    return this.__modal({
+      calLink,
+      enrichedConfig,
+      calOrigin,
+      isPrerendering,
+      stateData,
+      uid,
+      containerEl,
+    });
+  }
+
+  async __modal({
+    uid,
+    calLink,
+    enrichedConfig,
+    calOrigin,
+    isPrerendering,
```

Comment:
___modal_ does the actual job now, but called through `modal` which prepares the data to pass to it.